vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java

```diff
@@ -0,0 +1,105 @@
+package edu.asu.diging.vspace.core.services.impl;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+import edu.asu.diging.vspace.core.data.SequenceRepository;
+import edu.asu.diging.vspace.core.data.SlideRepository;
+import edu.asu.diging.vspace.core.factory.impl.SequenceFactory;
+import edu.asu.diging.vspace.core.model.ISequence;
+import edu.asu.diging.vspace.core.model.ISlide;
+import edu.asu.diging.vspace.core.model.impl.Module;
+import edu.asu.diging.vspace.core.model.impl.Sequence;
+import edu.asu.diging.vspace.core.model.impl.Slide;
+import edu.asu.diging.vspace.web.staff.forms.SequenceForm;
+
+public class SequenceManagerTest {
+
+    @Mock
+    private SequenceRepository mockSequenceRepo;
+
+    @Mock
+    private ModuleManager moduleManager;
+    
+    @Mock
+    private SlideManager slideManager;
+    
+    @Mock
+    private SlideRepository slideRepo;
+    
+    @Mock
+    private SequenceFactory sequenceFactory;
+    
+    @InjectMocks
+    private SequenceManager sequenceManagerToTest = new SequenceManager();
+    
+    @Before
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void test_getSequence_idExists() {
+        Sequence newSequence = new Sequence();
+        newSequence.setId("sequence1");
+        Optional<Sequence> sequence = Optional.of(newSequence);
+        Mockito.when(mockSequenceRepo.findById(newSequence.getId())).thenReturn(sequence);
+
+        ISequence actualSequence = sequenceManagerToTest.getSequence(newSequence.getId());
+        Assert.assertEquals(newSequence.getId(), actualSequence.getId());
+        Assert.assertEquals(newSequence.getName(), actualSequence.getName());
+    }
+    
+    @Test
+    public void test_getSequence_idNotExists() throws Exception {
+        Mockito.when(mockSequenceRepo.findById("a non existing id string")).thenReturn(Optional.empty());
+        Assert.assertNull(sequenceManagerToTest.getSequence("a non existing id string"));
+    }
+    
+    @Test
+    public void test_storeSequence_success() {        
+        Module module = new Module();
+        module.setId("module1");
+             
+        SequenceForm sequenceForm = new SequenceForm();
+        sequenceForm.setName("seq1");
+        sequenceForm.setDescription("sample description");            
+        List<String> slideIds = Arrays.asList("slide1", "slide2");
+        sequenceForm.setOrderedSlides(slideIds);
+        
+        List<ISlide> slidesList = Arrays.asList(new Slide(), new Slide());
+        slidesList.get(0).setId("slide1");
+        slidesList.get(1).setId("slide2");
+            
+        Sequence newSequence = new Sequence();
+        newSequence.setId("seq01");
```

Comment:
when ```storeSequence``` is called, the sequence does not have an id yet; the returned sequence has one (that's how you know it was stored). So  ```newSequence``` should not have an id. The sequence object returned by the repo, however, has.

Reply:
but we are the ones who are telling mockito what to return when save method is hit on the repo right? it is returning newSequence object and Assert.assertNotNull(actualSequence.getId()); is failing as it returns null.

Reply:
but let's say someone changes the code so that it checks if the ID is set after the factory is called and then returns null because it assumes the object is not a new object, then your test would fail although the code would still be correct. Your test case needs to mock what is actually happening (id = null before save, != null after save). Although, you probably want to test that the id the returned object has is the id that the save method returns, not a different one. Otherwise, if someone would change the code to set the id manually after save is called, your test would pass although the code would be wrong.

Reply:
the easiest would probably to create two sequence objects, one returned by the factory, one returned by the repo. the other option is to use ```doAnswer``` as described here (with actually exactly our case): https://wahz.blogspot.com/2012/10/mockito-modifying-parameters.html